source/poggles/program.cpp

```diff
@@ -5,13 +5,61 @@
 
 #include "poggles/gl_function.h"
 
+auto poggles::check_link_success(program_id identifier) -> bool
+{
+  GLint success = -1;
+
+  // check for link errors
+  glGetProgramiv(identifier, GL_LINK_STATUS, &success);
+  if (success == 0) {
+    GLint log_length = -1;
+    glGetProgramiv(identifier, GL_INFO_LOG_LENGTH, &log_length);
+    std::vector<char> log(static_cast<size_t>(log_length));
+    glGetProgramInfoLog(identifier, log_length, nullptr, log.data());
+
+    std::cout << "[PROGRAM] linking failed: " << log.data();
+
+    return false;
+  }
+
+  std::cout << "[PROGRAM] successfully compiled and linked program"
+            << std::endl;
+  ;
+  return true;
+}
+
+auto poggles::compileProgram(
+    program_id program,
+    std::initializer_list<std::pair<GLenum, std::string>> const& shaderFiles)
+    -> bool
+{
+  std::vector<shader_handle> shaders;
+  bool status = true;
+
+  for (auto [type, filename] : shaderFiles) {
+    shaders.push_back(std::move(shader_handle(type)));
+    status &= compileShader(shaders.back().value(), filename);
+    gl::attachShader(program, shaders.back().value());
+    // shader_handle should be good to go out of scope after being attached
+  }
+
+  glLinkProgram(program);
+
+  return status & poggles::check_link_success(program);
+}
+
+poggles::program::program() {}
+
 poggles::program::program(std::filesystem::path const& vertex_path,
```

Comment:
This version of the ctor seems like it's no longer needed, to be honest. It could just as easily take an initializer list of shader programs, like the `compileProgram` function does.  Or we could remove it and maybe even this class altogether?

Reply:
My main reason for having it was because I didn't want the compile program return boolean to be hidden in a constructor. That said, I want to delete the program class imminently, the only thing stopping me from doing it in this PR is that there will be cascading effects from it. Once the givderender PR is made and merged in I want to make a PR for that right away. I'll just do the initializer list function version for now, and remove the class in the next PR